client/src/app/shared/services/collaboration.service.ts

```diff
@@ -0,0 +1,147 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { CollaborationData, WebsocketService } from '../network/websocket.service';
+import { OverlayService } from 'src/app/overlay.service';
+import { AllocationsService } from '../data/allocations.service';
+import { ConstraintsService } from '../data/constraints.service';
+import { LockedStudentsService } from '../data/locked-students.service';
+import { StompSubscription } from '@stomp/stompjs';
+import { ConfirmationOverlayComponent } from 'src/app/components/confirmation-overlay/confirmation-overlay.component';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CollaborationService implements OnDestroy {
+  private discoverySubscription?: StompSubscription;
+  private subscriptions: StompSubscription[] = [];
+
+  constructor(
+    private websocketService: WebsocketService,
+    private overlayService: OverlayService,
+    private allocationsService: AllocationsService,
+    private constraintsService: ConstraintsService,
+    private lockedStudentsService: LockedStudentsService
+  ) {}
+
+  ngOnDestroy(): void {
+    for (const subscription of this.subscriptions || []) {
+      subscription.unsubscribe();
+    }
+    this.discoverySubscription?.unsubscribe();
+  }
+
+  private async discover(courseIterationId: string): Promise<CollaborationData> {
+    this.discoverySubscription?.unsubscribe();
+
+    return new Promise(async (resolve, reject) => {
```

Comment:
Promise executor functions should not be async. I think you added it because you wanna await the websocketService-Fucntion. Can you refactor the function.